### PR TITLE
[webOS] Acb (webOS 4.x) support

### PR DESCRIPTION
**Pull request by mariotaku**
## Description

On webOS 4.0, MediaPlayerAPIs has to be used with AcbAPI instead of exported window. This PR adds feature to use that when  exported window feature is not available.

## Motivation and context
To make Kodi support webOS 4.x

## How has this been tested?
Compiled app has been tested on webOS 5 TV and webOS 4 TV, both can perform video playback without issue.

## What is the effect on users?
webOS 4 users can now play videos

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by sundermann** [comment]
Thank you for polishing up the work on libacb!

I do see a small problem during packaging:
```
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libatomic.so.1
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libatomic.so.1.2.0
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libcrypt.so.1
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libcrypt-2.12.2.so
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libdrm.so.2
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libdrm.so.2.4.0
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libshairplay.so.0
-- Installing: /home/jenkins/jenkins-root/workspace/webos-dockerlinux_webosgles/build/tools/webOS/packaging/lib/libshairplay.so.0.0.0
```

The shim library `libacb` from `webos-userland` is not installed making kodi fail to launch on webOS 5+. A bit more context on this matter: webOS 5+ uses a different mechanism and thus does not bundle the deprecated libacb anymore. On webOS 5+ no acb functions are actually called so bundling the shim libraries from `webos-userland` is not a problem. This works because the jailed environment is set up in a way that system libraries will take precedence over any custom bundled ones. So bundling `libacb` is needed to make kodi able to resolve the library/functions on startup. Just add `libacb` [here](https://github.com/webosbrew/xbmc/blob/feature/webOS/cmake/scripts/webos/Install.cmake#L31)

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp [hunk 1]
```diff
@@ -556,3 +579,91 @@ void CDVDVideoCodecStarfish::PlayerCallback(const int32_t type,
...
+{
+  if (acbId)
+    AcbAPI_destroy(acbId);
+}
+
+bool StarfishResourceAcb::PostLoad(CDVDStreamInfo& hints)
+{
+  if (!acbId)
+    return false;
+  AcbAPI_setDisplayWindow(acbId, 0, 0, hints.width, hints.height, true, nullptr);
```
Comment (sundermann) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp:628]: Unfortunately, this has to happen in `CRendererStarfish`. I'd suggest adding a field for `acbId` in the `StarfishVideoBuffer`. Then calling the function in `ManageRenderArea`. Also, videos may not always run fullscreen. For instance, if you display a SBS 3D video as 2D. There is some logic inside the renderer that calculates the target display coordinates. There's another catch here though. The target coordinates can be negative which the API will not accept. You can test this when a video partly goes offscreen, i.e. when using a zoom option > 1.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp [hunk 2]
```diff
@@ -353,7 +364,11 @@ bool CDVDVideoCodecStarfish::AddData(const DemuxPacket& packet)
     std::string result = m_starfishMediaAPI->Feed(json.c_str());
 
     if (result.find("Ok") != std::string::npos)
+    {
+      if (startPlaying)
+        m_starfishResource->StartPlaying();
```
Comment (sundermann) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp:369]: Is this here placed correctly? I could imagine that it'd be better placed when a `PF_EVENT_TYPE_STR_STATE_UPDATE__LOADCOMPLETED` or `PF_EVENT_TYPE_STR_STATE_UPDATE__PLAYING` is received. Maybe also call `AcbAPI_setState(acbId, APPSTATE_FOREGROUND, PLAYSTATE_PAUSED, NULL);` when `PF_EVENT_TYPE_STR_STATE_UPDATE__PAUSED` is received.

**Review by sundermann — CHANGES_REQUESTED** [review]

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp [hunk 3]
```diff
@@ -32,6 +51,21 @@ CBaseRenderer* CRendererStarfish::Create(CVideoBuffer* buffer)
 
 bool CRendererStarfish::Configure(const VideoPicture& picture, float fps, unsigned int orientation)
 {
+  if (m_acbId)
+  {
+    if (!AcbAPI_initialize(m_acbId, PLAYER_TYPE_MSE, getenv("APPID"), AcbCallback))
```
Comment (sundermann) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp:56]: This is probably also better happening in the decoder

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp [hunk 4]
```diff
@@ -156,3 +209,14 @@ void CRendererStarfish::RenderUpdate(
     return;
   }
 }
+
+static void AcbCallback(
+    long acbId, long taskId, long eventType, long appState, long playState, const char* reply)
+{
+  (void)acbId;
```
[the remark this answers is not in the document]
Comment (sundermann) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp:217]: Also why do you need to cast all of these to `void`? Avoiding unused parameter warnings?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp [hunk 5]
```diff
@@ -524,6 +527,7 @@ void CDVDVideoCodecStarfish::PlayerCallback(const int32_t type,
   CLog::LogF(LOGDEBUG, "CDVDVideoCodecStarfish: type: {}, numValue: {}, strValue: {}", type,
              numValue, logstr);
 
+  auto buffer = dynamic_cast<CStarfishVideoBuffer*>(m_videobuffer.videoBuffer);
```
Comment (sundermann) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp:530]:
```suggestion
  auto buffer = static_cast<CStarfishVideoBuffer*>(m_videobuffer.videoBuffer);
```

**Review by sundermann — CHANGES_REQUESTED** [review]

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mariotaku) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp:217]: Yes exactly. I'll ad some logging instead.

**Comment by Danar435** [comment]
You have my sincerest gratitude

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp [hunk 6]
```diff
@@ -533,17 +561,42 @@ void CDVDVideoCodecStarfish::PlayerCallback(const int32_t type,
     case PF_EVENT_TYPE_STR_RESOURCE_INFO:
       m_newFrame = true;
       break;
+    case PF_EVENT_TYPE_STR_VIDEO_INFO:
+      AcbAPI_setMediaVideoData(m_acbId, logstr.c_str());
```
Comment (yol) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecStarfish.cpp:565]: `if (m_acbId)` missing?

**Review by sundermann — CHANGES_REQUESTED** [review]

**Review by sundermann — APPROVED** [review]

**Comment by sundermann** [comment]
Can you squash the fixups?

**Comment by sundermann** [comment]
Jenkins build this please

**Comment by DannAberHardt** [comment]
Tested on LG C8 and picture and sound is working :-)
Audio Passthrough does not work on arc for me.

**Comment by sundermann** [comment]
It is possible the passthrough starfish sink needs a similar acb integration as the video decoder.

**Comment by sundermann** [comment]
@DannAberHardt Maybe you can test #23866 build once it's available?

**Pull request merged**